cmd/note_common.go

```diff
@@ -166,42 +165,64 @@ func createCommitComments(project string, mrID int, commit string, body string,
 			}
 		} else if strings.HasPrefix(scanner.Text(), "|") {
 			if comments != "" {
-				createCommitNote(project, mrID, commit, newfile, oldfile, oldLineNum, newLineNum, comments, block)
+				createCommitNote(project, mrID, commit, newfile, oldfile, linetype, oldLineNum, newLineNum, comments, block)
```

Comment:
> Frustratingly I cannot get the code to work for context lines at the end of the diff :/  I've tried everything I could think of, including adding linePosition entries and forcing the line_code.  I still get the same error.

Anyway here is a minimal fix. I had to insert a

```go
	fmt.Printf("%#v\n", position)
```

which showed that we send the wrong "old" line for context lines.
Would be nice if `lab` could print `curl` commands, then it'd be easier to check with other tools and the API docs.

```diff
diff --git a/cmd/note_common.go b/cmd/note_common.go
index 6f6bd14..4bf44e9 100644
--- a/cmd/note_common.go
+++ b/cmd/note_common.go
@@ -94,14 +94,9 @@ func noteRunFn(cmd *cobra.Command, args []string) {
 	createNote(rn, isMR, int(idNum), msgs, filename, linebreak, commit, true)
 }
 
-func createCommitNote(rn string, mrID int, sha string, newFile string, oldFile string, linetype string, oldline int, newline int, comment string, block bool) {
-	line := oldline
-	if oldline == -1 {
-		line = newline
-	}
-
+func createCommitNote(rn string, mrID int, sha string, newFile string, oldFile string, linetype string, oldLine int, newLine int, comment string, block bool) {
 	if block {
-		webURL, err := lab.CreateMergeRequestCommitDiscussion(rn, mrID, sha, newFile, oldFile, line, linetype, comment)
+		webURL, err := lab.CreateMergeRequestCommitDiscussion(rn, mrID, sha, newFile, oldFile, oldLine, newLine, linetype, comment)
 		if err != nil {
 			log.Fatal(err)
 		}
@@ -109,7 +104,7 @@ func createCommitNote(rn string, mrID int, sha string, newFile string, oldFile s
 		return
 	}
 
-	webURL, err := lab.CreateCommitComment(rn, sha, newFile, oldFile, line, linetype, comment)
+	webURL, err := lab.CreateCommitComment(rn, sha, newFile, oldFile, oldLine, newLine, linetype, comment)
 	if err != nil {
 		log.Fatal(err)
 	}
@@ -190,6 +185,7 @@ func createCommitComments(project string, mrID int, commit string, body string,
 			val2, err := strconv.Atoi(f[2])
 			if err == nil { // f[2] is a number
 				if len(f) > 3 { // f[3] exists
+					oldLineNum = val1
 					newLineNum = val2
 					// f[3] starts with a +
 					if strings.HasPrefix(f[3], "+") {
diff --git a/internal/gitlab/gitlab.go b/internal/gitlab/gitlab.go
index 3793c3f..d6696bc 100644
--- a/internal/gitlab/gitlab.go
+++ b/internal/gitlab/gitlab.go
@@ -1695,7 +1695,7 @@ func GetCommitDiff(projID string, sha string) ([]*gitlab.Diff, error) {
 	return diffs, nil
 }
 
-func CreateCommitComment(projID string, sha string, newFile string, oldFile string, line int, linetype string, comment string) (string, error) {
+func CreateCommitComment(projID string, sha string, newFile string, oldFile string, oldLine int, newLine int, linetype string, comment string) (string, error) {
 	// Ideally want to use lab.Commits.PostCommitComment, however,
 	// that API only support comments on linetype=new.
 	//
@@ -1721,15 +1721,15 @@ func CreateCommitComment(projID string, sha string, newFile string, oldFile stri
 	switch linetype {
 	case "new":
 		position.NewPath = newFile
-		position.NewLine = line
+		position.NewLine = newLine
 	case "old":
 		position.OldPath = oldFile
-		position.OldLine = line
+		position.OldLine = oldLine
 	case "context":
 		position.NewPath = newFile
-		position.NewLine = line
+		position.NewLine = newLine
 		position.OldPath = oldFile
-		position.OldLine = line
+		position.OldLine = oldLine
 	}
 
 	opt := &gitlab.CreateCommitDiscussionOptions{
@@ -1745,7 +1745,7 @@ func CreateCommitComment(projID string, sha string, newFile string, oldFile stri
 	return fmt.Sprintf("%s#note_%d", commitInfo.WebURL, commitDiscussion.Notes[0].ID), nil
 }
 
-func CreateMergeRequestCommitDiscussion(projID string, id int, sha string, newFile string, oldFile string, line int, linetype string, comment string) (string, error) {
+func CreateMergeRequestCommitDiscussion(projID string, id int, sha string, newFile string, oldFile string, oldLine int, newLine int, linetype string, comment string) (string, error) {
 	commitInfo, err := GetCommit(projID, sha)
 	if err != nil {
 		fmt.Printf("Could not get diff for commit %s.\n", sha)
@@ -1768,12 +1768,12 @@ func CreateMergeRequestCommitDiscussion(projID string, id int, sha string, newFi
 
 	switch linetype {
 	case "new":
-		position.NewLine = line
+		position.NewLine = newLine
 	case "old":
-		position.OldLine = line
+		position.OldLine = oldLine
 	case "context":
-		position.NewLine = line
-		position.OldLine = line
+		position.NewLine = newLine
+		position.OldLine = oldLine
 	}
 
 	opt := &gitlab.CreateMergeRequestDiscussionOptions{
```

Reply:
Yeah, that's what I've been doing.  But even when I do set a "valid" position at the end of the commit, the command to comment returns an error:
2021/12/10 19:11:31 ERROR: note_common.go:114: POST https://gitlab.com/api/v4/projects/redhat/rhel/src/kernel/kernel-test/repository/commits/abf523e2e10da39b15f522378838bf4d39604e15/discussions: 400 {message: 400 Bad request - Note {:line_code=>["can't be blank", "must be a valid line code"]}}

Reply:
I'll look into seeing if go-gitlab has a debug option.  IIRC it does but I've never used it.

Reply:
I didn't get any errors with this diff applied, do you have a reproducible example?